### DISCUSSION TITLE
ffi w/ webassembly

### DISCUSSION THREAD

**Discussion by unicomp21** [dart-lang/ffi#114] [Q&A] [answered]
If I wanted to create ffi bindings for filament webassembly/webgl, is there an example somewhere which could help me get started?

https://github.com/google/filament/blob/24b88219fa6148b8004f230b377f163e6f184d65/build/web/build.sh#L35

**Comment by dcharkes** [comment, marked as answer]
Could you elaborate a bit more? I'm assuming you want to compile a native library to WebAssembly and use that on the web with the dart2js compiler?

You would use https://dart.dev/web/js-interop to talk to a native library compiled to WebAssembly, not `dart:ffi`+`package:ffi`.

We would want to like explore creating a library that would target both JS-Interop and Dart FFI, but we haven't started on that yet.

**Reply by unicomp21** [reply]
@dcharkes thanks for setting me straight, I mistakenly thought dart ffi worked with webassembly in the browser.

https://github.com/dart-lang/wasm/search?q=dart%3Affi

the intention was to program against the filament typescript api using dart, but use ffi bindings instead, and run it all in the browser.  filament has a webgl/browser build option.

i'm also thinking there might be performance arguments for not using the js-interop support.

looks like i'll have to settle for js-interop atm.

**Reply by dcharkes** [reply]
I have filed https://github.com/dart-lang/sdk/issues/46690 to track ergonomics of using both JS-interop with WASM _and_ `dart:ffi` with the same native library.
